test-data/unit/check-union-or-syntax.test

```diff
@@ -196,3 +196,47 @@ def f(x: Union[int, str, None]) -> None:
     else:
         reveal_type(x)  # N: Revealed type is "None"
 [builtins fixtures/isinstance.pyi]
+
+[case testUnionOrSyntaxMetaclassOverride]
+# flags: --python-version 3.10
+import types
+
+class Meta(type):
+    def __or__(self, other) -> types.UnionType: pass
+
+class F(metaclass=Meta): pass
+
+reveal_type(F | F)  # N: Revealed type is "types.UnionType"
```

Comment:
I haven't personally used this trick yet, so I am thinking 🤔 
Should not this be just `F`? Or do we want this special treatment of `types.UnionType`?

Reply:
I think there is confusion over `Union` and `UnionType` here (and imo, it is confusing because unions are treated so specially). `UnionType` is the runtime type of a `Union` object:
```
X = str | int
print(type(X)) # <class 'types.UnionType'>
```
So, the type of X is always UnionType (it's worth noting that in previous version of mypy this was just `builtins.object`).

---

This is perhaps easier to grasp given:
```
class A: ...
class A: ...
reveal_type(A | B)  # types.UnionType
f: F
reveal_type(f) # Union[test2.A, test2.A]
```

Now, the second case not reducing is almost certainly a bug, but I think that is unrelated to this specific issue (it's also present in master) since it doesn't matter if its a `TypeAlias` or not (just if it was a PEP 604 union or not).